components/raftstore-v2/src/operation/raft_log_gc.rs

```diff
@@ -0,0 +1,154 @@
+            }
+            if let Some(last_heartbeat) = self.peer_heartbeats.get(peer_id) {
+                if *last_heartbeat > cache_alive_limit {
+                    if alive_cache_idx > p.matched && p.matched >= truncated_idx {
+                        alive_cache_idx = p.matched;
+                    } else if p.matched == 0 {
+                        // the new peer is still applying snapshot, do not compact cache now
+                        alive_cache_idx = 0;
+                    }
+                }
+            }
+        }
+
+        // When an election happened or a new peer is added, replicated_idx can be 0.
+        if replicated_idx > 0 {
+            assert!(
+                last_idx >= replicated_idx,
+                "expect last index {} >= replicated index {}",
+                last_idx,
+                replicated_idx
+            );
+        }
+
+        // leader may call `get_term()` on the latest replicated index, so compact
+        // entries before `alive_cache_idx` instead of `alive_cache_idx + 1`.
+        self.entry_storage_mut()
+            .compact_entry_cache(std::cmp::min(alive_cache_idx, applied_idx + 1));
+
+        let mut compact_idx = if (applied_idx > first_idx
+            && applied_idx - first_idx >= store_ctx.cfg.raft_log_gc_count_limit())
+            || (self.raft_log_size_hint >= store_ctx.cfg.raft_log_gc_size_limit().0)
+        {
+            std::cmp::max(first_idx + (last_idx - first_idx) / 2, replicated_idx)
+        } else if replicated_idx < first_idx || last_idx - first_idx < 3 {
+            return;
+        } else if replicated_idx - first_idx < store_ctx.cfg.raft_log_gc_threshold
+            && self.skip_gc_raft_log_ticks < store_ctx.cfg.raft_log_reserve_max_ticks
+        {
+            self.skip_gc_raft_log_ticks += 1;
+            return;
+        } else {
+            replicated_idx
+        };
+        assert!(compact_idx >= first_idx);
+        // Have no idea why subtract 1 here, but original code did this by magic.
+        compact_idx -= 1;
+        if compact_idx < first_idx {
+            return;
+        }
+
+        // Create a compact log request and notify directly.
+        let region_id = self.region().get_id();
+        let peer = self.peer().clone();
+        // TODO: move this into a function
+        let term = self.raft_group().raft.raft_log.term(compact_idx).unwrap();
+
+        let mut req = self.new_admin_request();
+        let mut admin = AdminRequest::default();
+        admin.set_cmd_type(AdminCmdType::CompactLog);
+        admin.mut_compact_log().set_compact_index(compact_idx);
+        admin.mut_compact_log().set_compact_term(term);
+        req.set_admin_request(admin);
+
+        let (msg, _) = PeerMsg::raft_command(req);
+        if let Err(e) = store_ctx.router.send(self.region_id(), msg) {
+            error!(self.logger, "send compact log request failed");
+        }
+        // self.propose_command(store_ctx, req);
```

Comment:
`// self.propose_command(store_ctx, req);`
is it needed? If not, the comment can be deleted

Reply:
Because I'm hesitating on the implementation. It's just very weird to use router to send message to the current peer. Alternatively I can directly call `on_admin_command` here, but I'm not sure if it's the intended usage of it. cc @BusyJay

Reply:
Yes, you can call on_admin_command directly.